File: 프로그래머스/lv0/120956. 옹알이 （1）/옹알이 （1）.java
class Solution {
    public int solution(String[] babbling) {
        int count = 0;
        for (String str : babbling) {
            if (canPronounce(str)) {
                count++;
            }
        }
        return count;
    }
    private boolean canPronounce(String str) {
        String[] words = {"aya", "ye", "woo", "ma"};
        for (String word : words) {
            if (str.startsWith(word)) {
                String remaining = str.substring(word.length());
                if (remaining.isEmpty()) {
                    return true;
                }
                //재귀
                if (canPronounce(remaining)) {
                    return true;
                }
            }
        }
        return false;
    }

}

File: 프로그래머스/lv0/120956. 옹알이 （1）/옹알이 （1）.js
function solution(babbling) {
    let count = 0;
    for (let str of babbling) {
        if (canPronounce(str)) {
            count++;
        }
    }
    return count;
}

function canPronounce(str) {
    let words = ["aya", "ye", "woo", "ma"];
    for (let word of words) {
        if (str.startsWith(word)) {
            let remaining = str.substring(word.length);
            if (remaining === "") {
                return true;
            }
            if (canPronounce(remaining)) {
                return true;
            }
        }
    }
    return false;
}
